x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExportBulk.java

```diff
@@ -77,12 +75,11 @@ public void doAdd(Collection<MonitoringDoc> docs) throws ExportException {
             if (docs != null && docs.isEmpty() == false) {
                 try (BytesStreamOutput payload = new BytesStreamOutput()) {
                     for (MonitoringDoc monitoringDoc : docs) {
-                        // any failure caused by an individual doc will be written as an empty byte[], thus not impacting the rest
-                        payload.write(toBulkBytes(monitoringDoc));
+                        writeDocument(monitoringDoc, payload);
                     }
 
                     // store the payload until we flush
-                    this.payload = BytesReference.toBytes(payload.bytes());
+                    this.payload = payload.bytes();
```

Comment:
Now that we are not copying bytes, is this still thread safe ? e.g. if one thread is calling doAdd, but another thread is calling doFlush , couldn't this line swap out the reference that flush is using ? (i think resulting in missing documents)

Reply:
Looking at `org.elasticsearch.xpack.monitoring.exporter.ExportBulk#flush`  it seems that calling flush is only allowed once. So if we were adding and flushing concurrently anywhere that would be broken already anyway?

```
    public void flush(ActionListener<Void> listener) {
        if (closed.compareAndSet(false, true)) {
            doFlush(listener);
        } else {
            listener.onResponse(null);
        }
    }
```

Reply:
I see now, the doAdd (in practice) is also guarded by this.